RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletHeterogeneousEDProducer.cc

```diff
@@ -0,0 +1,206 @@
+#include "DataFormats/Common/interface/Handle.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/PluginManager/interface/ModuleDef.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/RunningAverage.h"
+#include "HeterogeneousCore/CUDACore/interface/GPUCuda.h"
+#include "HeterogeneousCore/CUDAServices/interface/CUDAService.h"
+#include "HeterogeneousCore/Producer/interface/HeterogeneousEDProducer.h"
+
+#include "CAHitQuadrupletGenerator.h"
+#include "CAHitQuadrupletGeneratorGPU.h"
+#include "RecoPixelVertexing/PixelTriplets/interface/OrderedHitSeeds.h"
+#include "RecoTracker/TkHitPairs/interface/IntermediateHitDoublets.h"
+#include "RecoTracker/TkHitPairs/interface/RegionsSeedingHitSets.h"
+
+namespace {
+void fillNtuplets(RegionsSeedingHitSets::RegionFiller &seedingHitSetsFiller,
+                  const OrderedHitSeeds &quadruplets) {
+  for (const auto &quad : quadruplets) {
+    seedingHitSetsFiller.emplace_back(quad[0], quad[1], quad[2], quad[3]);
+  }
+}
+} // namespace
+
+class CAHitNtupletHeterogeneousEDProducer
+    : public HeterogeneousEDProducer<heterogeneous::HeterogeneousDevices<
+          heterogeneous::GPUCuda, heterogeneous::CPU>> {
+public:
+  CAHitNtupletHeterogeneousEDProducer(const edm::ParameterSet &iConfig);
+  ~CAHitNtupletHeterogeneousEDProducer() = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions);
+  void beginStreamGPUCuda(edm::StreamID streamId,
+                          cuda::stream_t<> &cudaStream) override;
+
+  void acquireGPUCuda(const edm::HeterogeneousEvent &iEvent,
+                      const edm::EventSetup &iSetup,
+                      cuda::stream_t<> &cudaStream) override;
+  void produceGPUCuda(edm::HeterogeneousEvent &iEvent,
+                      const edm::EventSetup &iSetup,
+                      cuda::stream_t<> &cudaStream) override;
+  void produceCPU(edm::HeterogeneousEvent &iEvent,
+                  const edm::EventSetup &iSetup) override;
+
+private:
+  edm::EDGetTokenT<IntermediateHitDoublets> doubletToken_;
+
+  edm::RunningAverage localRA_;
+  CAHitQuadrupletGeneratorGPU GPUGenerator_;
+  CAHitQuadrupletGenerator CPUGenerator_;
+
+  bool emptyRegionDoublets = false;
+  std::unique_ptr<RegionsSeedingHitSets> seedingHitSets;
+  std::vector<OrderedHitSeeds> ntuplets;
```

Comment:
please update the member variable names to match the coding rules, i.e. add a trailing `_`